src/main/java/com/direwolf20/buildinggadgets/common/entities/BlockBuildEntity.java

```diff
@@ -21,62 +20,105 @@
 import java.util.Optional;
 
 public class BlockBuildEntity extends EntityBase {
-    private static final DataParameter<Integer> toolMode = EntityDataManager.<Integer>createKey(BlockBuildEntity.class, DataSerializers.VARINT);
+
+    public static enum Mode {
+        PLACE("place") {
+            @Override
+            public void onBuilderEntityDespawn(BlockBuildEntity builder) {
+                World world = builder.world;
+                BlockPos targetPos = builder.targetPos;
+                IBlockState targetBlock = builder.setBlock;
+                if (builder.isUsingPaste()) {
+                    world.setBlockState(targetPos, BGBlocks.constructionBlock.getDefaultState());
+                    TileEntity te = world.getTileEntity(targetPos);
+                    if (te instanceof ConstructionBlockTileEntity) {
+                        ((ConstructionBlockTileEntity) te).setBlockState(targetBlock, targetBlock);
+                    }
+                    world.spawnEntity(new ConstructionBlockEntity(world, targetPos, false));
+                } else {
+                    world.setBlockState(targetPos, targetBlock);
+                    BlockPos upPos = targetPos.up();
+                    world.getBlockState(targetPos).neighborChanged(world, targetPos, world.getBlockState(upPos).getBlock(), upPos);
+                }
+            }
+        },
+        REMOVE("remove") {
+            @Override
+            public void onBuilderEntityDespawn(BlockBuildEntity builder) {
+                builder.world.setBlockState(builder.targetPos, Blocks.AIR.getDefaultState());
+            }
+        },
+        REPLACE("replace") {
+            @Override
+            public void onBuilderEntityDespawn(BlockBuildEntity builder) {
+                World world = builder.world;
+                world.spawnEntity(new BlockBuildEntity(world, builder.targetPos, builder.spawnedBy, builder.originalSetBlock, PLACE, builder.isUsingPaste()));
+            }
+        };
+
+        private final String id;
```

Comment:
Could you use int or byte id's for efficiency reasons? There's no need the DataManager syncs a whole string over the network if you only need to know which mode...

Reply:
```suggestion
        private static final Byte2ObjectMap<Mode> reverseMap = new Byte2ObjectHashMap<>();
        private final byte id;
```

Reply:
Then I would rather use `ordinal()` -- consistency would not be a problem, as there is almost no need for extension.

Reply:
There might be need for it in the future, but if you put a comment there, not to change the order for this reason it should be fine in an internal class...